### PR TITLE
refactor!: isolate arrow code

### LINKED ISSUE #237: Isolate `arrow` code
# Background and Motivation

Currently, `arrow` is one of the elements that must be disabled to enable `no_std` support. Additionally, not every use-case requires arrow. So, we have gated it behind a features flag (`feature = "arrow"`). However, there are many instances of `#[cfg(feature = "arrow")]` throughout the codebase, which leave the code cluttered.

# Changes Required

- [ ] Move all arrow code to a single (or a small number) of modules.
    - My suggestion would be creating a new `proof_of_sql::base::arrow` module and moving all code there. However, this may not be possible, so this is not a hard requirement.
    - The resulting code _**must**_ not introduce cyclic dependencies and should resolve any existing ones with arrow related code. This is one of the primary reasons for this cleanup: in order to more easily identify and resolve these.
- [ ] Update any legacy documentation referencing arrow to ensure that it is up to date.
- [ ] Bump the `arrow` crate to version 53.

### PR DESCRIPTION
**Pull request by varshith257**
Please be sure to look over the pull request guidelines here: https://github.com/spaceandtimelabs/sxt-proof-of-sql/blob/main/CONTRIBUTING.md#submit-pr.

# Please go through the following checklist
- [ ] The PR title and commit messages adhere to guidelines here: https://github.com/spaceandtimelabs/sxt-proof-of-sql/blob/main/CONTRIBUTING.md. In particular `!` is used if and only if at least one breaking change has been introduced.
- [ ] I have run the ci check script with `source scripts/run_ci_checks.sh`.

# Rationale for this change
Currently, arrow is one of the elements that must be disabled to enable no_std support. There are many instances of #[cfg(feature = "arrow")] throughout the codebase, which leaves the code cluttered. To make code readable and manageable  , arrow module is introduced.

Closes #237 
/claim #237 

# What changes are included in this PR?

- Moved all  [cfg(feature = "arrow")] related code to separate modules and cracked cyclic dependencies along the way without any massive disruption. 
- Bumped Arrow from 51 -> 53

# Are these changes tested?

Yes

### CODE REVIEW

**Comment by varshith257** [comment]
@JayWhite2357 Have a review in your free time.

File: crates/proof-of-sql/src/base/arrow/mod.rs [hunk 1]
```diff
@@ -0,0 +1,46 @@
+//! This module provides conversions and utilities for working with Arrow data structures.
+
+#[cfg(feature = "arrow")]
```
Comment (iajoiner) [inline crates/proof-of-sql/src/base/arrow/mod.rs:3]: You probably want `#[cfg(feature = "arrow")]` on `pub mod arrow` in `base/mod.rs` instead and remove all of them here

File: crates/proof-of-sql/src/base/mod.rs [hunk 2]
```diff
@@ -1,5 +1,6 @@
 //! This module contains basic shared functionalities of the library.
 /// TODO: add docs
+pub mod arrow;
```
Comment (iajoiner) [inline crates/proof-of-sql/src/base/mod.rs:3]:
```rust
#[cfg(feature = "arrow")]
```
Please put it above `pub mod arrow` and remove all instances of it within the `arrow` directory

File: crates/proof-of-sql/src/base/arrow/mod.rs (hunk 1 above)
Comment (varshith257) [reply crates/proof-of-sql/src/base/arrow/mod.rs:3]: ok

**Review by JayWhite2357 — CHANGES_REQUESTED** [review]
Looks pretty good so far. The main concern I have is around cyclic dependencies.

> The resulting code must not introduce cyclic dependencies and should resolve any existing ones with arrow related code. This is one of the primary reasons for this cleanup: in order to more easily identify and resolve these.

In particular, nothing in `base` should depend on anything outside of `base`.

I've added comments to this effect. If you resolve those along with @iajoiner's concerns, we should be good on this PR.

**Comment by varshith257** [comment]
@JayWhite2357 Will bump with the fixes

**Comment by varshith257** [comment]
@JayWhite2357 @iajoiner Can have a look at it here?
https://github.com/spaceandtimelabs/sxt-proof-of-sql/actions/runs/11417714964/job/31770330872#step:4:766

I am not sure what it making to fail

**Comment by melsonic** [comment]
> @JayWhite2357 @iajoiner Can have a look at it here? https://github.com/spaceandtimelabs/sxt-proof-of-sql/actions/runs/11417714964/job/31770330872#step:4:766
> 
> I am not sure what it making to fail

I believe It's because of being unable to find the implementations in `crates/proof-of-sql/src/base/arrow/arrow_query_result.rs`.

**Comment by JayWhite2357** [comment]
> I am not sure what it making to fail

@varshith257 

replace
```rust
RecordBatch::try_from(query_result).unwrap()
```
with
```rust
RecordBatch::try_from(query_result.table).unwrap()
```

This is caused because I asked you to remove this code:
```rust
impl<S: Scalar> TryFrom<QueryData<S>> for RecordBatch {
    type Error = ArrowError;

    fn try_from(value: QueryData<S>) -> Result<Self, Self::Error> {
        Self::try_from(value.table)
    }
}
```

**Review by iajoiner — DISMISSED** [review]
@varshith257 Really thanks for your great work!

**Comment by varshith257** [comment]
@JayWhite2357 @iajoiner PTAL

**Review by JayWhite2357 — APPROVED** [review]
LGTM

**Pull request merged**